src/PatternkitDrupalTwigLib.php

```diff
@@ -134,53 +146,93 @@ public function getEditor($subtype = NULL,
-      }
-      $file_path = $file->getPath();
-      $dirs = explode(DIRECTORY_SEPARATOR, $file_path);
-      // All JSON schema must be in an 'api' folder at this time.
-      // @todo Add support for custom setups.
-      // @todo Look at a standard for JSON Schema + JSON Sample data.
-      $num_dirs = count($dirs);
-      if ($num_dirs < 2
-        || array_pop($dirs) !== 'api') {
-        continue;
-      }
-      $file_ext = $file->getExtension();
-      if (!in_array(strtolower($file_ext), $filter, TRUE)) {
-        continue;
+    static $metadata;
+
+    // Use static pattern to avoid rebuilding multiple times per request.
+    if (is_null($metadata)) {
+
+      $it = new RecursiveDirectoryIterator($this->path);
+      $filter = ['json', 'twig'];
+      $metadata = [];
+      $components = [];
+
+      /** @var \SplFileInfo $file */
+      foreach (new RecursiveIteratorIterator($it) as $file) {
+        // Skip directories and non-files.
+        if (!$file->isFile()) {
+          continue;
+        }
+        $file_path = $file->getPath();
+
+        // Skip tests folders.
+        if (strpos($file_path, '/tests') !== FALSE) {
+          continue;
+        }
+
+        // Get the file extension for the file.
+        $file_ext = $file->getExtension();
+        if (!in_array(strtolower($file_ext), $filter, TRUE)) {
+          continue;
+        }
+
+        // We use file_basename as a unique key, it is required that the
+        // JSON and twig file share this basename.
+        $file_basename = $file->getBasename('.' . $file_ext);
+
+        // Build an array of all the filenames of interest, keyed by name.
+        $components[$file_basename][$file_ext] = $file_path;
       }
-      if ($file_contents = file_get_contents($file)) {
-        $pattern = $this->createPattern(json_decode($file_contents));
-        $file_basename = $file->getBasename('.json');
-        $subtype = "pk_$file_basename";
-        $pattern->subtype = $subtype;
-        $pattern->url = url("patternkit/ajax/$id/$subtype/schema");
-        $twig_file = $file_path
-          . DIRECTORY_SEPARATOR . $file_basename . '.twig';
-        if (file_exists($twig_file)) {
-          $pattern->filename = $twig_file;
-          $pattern->template = file_get_contents($twig_file);
+
+      foreach ($components as $module_name => $data) {
+        // If the component has a json file, create the pattern from it.
+        if (!empty($data['json']) && $file_contents = file_get_contents($data['json'])) {
+          $pattern = $this->createPattern(json_decode($file_contents));
+
+          $subtype = "pk_$module_name";
+          $pattern->subtype = $subtype;
+          // URL is redundant for the twig based components.
+          $pattern->url = $module_name;
+        }
+        else {
+          // Create the pattern from defaults.
+          $pattern = $this->createPattern(
+            (object) [
+              '$schema'    => 'http =>//json-schema.org/draft-04/schema#',
```

Comment:
Should these defaults be in the `PatternkitPattern` class?

Reply:
That whole section needs a refactor. The problem was that some folders have a .json, but no .twig, and some vice versa. This is particularly problematic because of the nested dependencies.  no_html.twig was a recurring one in particular.

Reply:
Gotcha. I'll put it into the roadmap, thanks!